include/triton/core/tritonbackend.h

```diff
@@ -711,6 +743,67 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_StateBuffer(
     TRITONBACKEND_State* state, void** buffer, const uint64_t buffer_byte_size,
     TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id);
 
+/// Get the buffer attributes associated with the given state buffer.
+/// The returned 'buffer_attributes' is owned by the state and so should not be
+/// modified or freed by the caller. The lifetime of the 'buffer_attributes'
+/// matches that of the state.
+///
+/// \param state The state.
+/// \param buffer_attributes Returns the buffer attributes for the given state.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_StateBufferAttributes(
+    TRITONBACKEND_State* state,
+    TRITONBACKEND_BufferAttributes* buffer_attributes);
+
+///
+/// TRITONBACKEND_BufferAttributes
+///
+/// Object representing the attributes associated with an input, output, or
+/// state buffer. The API supports accessing the memory_type, memory_type_id,
+/// and CUDA IPC handle.
+///
+
+/// Get the memory type associated with the buffer.
+///
+/// \param buffer_attributes The buffer attributes object.
+/// \param memory_type Returns the memory type of the buffer.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_BufferAttributesMemoryType(
+    TRITONBACKEND_BufferAttributes* buffer_attributes,
+    TRITONSERVER_MemoryType* memory_type);
+
+/// Get the memory type id associated with the buffer.
+///
+/// \param buffer_attributes The buffer attributes object.
+/// \param memory_type_id Returns the memory type id of the buffer.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_BufferAttributesMemoryTypeId(
+    TRITONBACKEND_BufferAttributes* buffer_attributes, int64_t* memory_type_id);
+
+/// Get the byte size associated with the buffer.
+///
+/// \param buffer_attributes The buffer attributes object.
+/// \param byte_size Returns the byte size of the buffer.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_BufferAttributesByteSize(
+    TRITONBACKEND_BufferAttributes* buffer_attributes, size_t* byte_size);
+
+/// Get the cudaIpcHandle associated with the buffer. If the cudaIpcHandle does
+/// not exist for the given buffer, the handle will contain nullptr.  The
+/// returned 'handle' is owned by the 'buffer_attributes' and so should not be
+/// modified or freed by the caller. The lifetime of the 'handle' matches that
+/// of the 'buffer_attributes'.
+///
+/// \param buffer_attributes The buffer attributes object.
+/// \param handle Returns the memory type id of the buffer.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_BufferAttributesCudaIpcHandle(
+    TRITONBACKEND_BufferAttributes* buffer_attributes, void** handle);
```

Comment:
should the "handle" type just be the cuda IPC handle?

Reply:
I thought about this but this would add a dependency to the cuda runtime header file which may not be available in all circumstances (e.g. CPU only builds).